src/sage/schemes/hyperelliptic_curves/constructor.py

```diff
@@ -265,19 +279,22 @@ def is_pAdicField(x):
     fields = [
         ("FiniteField", is_FiniteField, HyperellipticCurve_finite_field),
         ("RationalField", is_RationalField, HyperellipticCurve_rational_field),
-        ("pAdicField", is_pAdicField, HyperellipticCurve_padic_field)]
-
-    if g in genus_classes:
-        superclass.append(genus_classes[g])
-        cls_name.append("g%s" % g)
+        ("pAdicField", is_pAdicField, HyperellipticCurve_padic_field),
+    ]
 
     for name, test, cls in fields:
         if test(R):
-            superclass.append(cls)
+            bases.append(cls)
             cls_name.append(name)
             break
 
+    # If no specialised child class was found, we simply use the
+    # generic class in the class construction
+    if not bases:
+        bases = [HyperellipticCurve_generic]
+
+    # Dynamically build a class from multiple inheritance. Note that
+    # all classes we slect from are children of HyperellipticCurve_generic
     class_name = "_".join(cls_name)
-    cls = dynamic_class(class_name, tuple(superclass),
-                        HyperellipticCurve_generic, doccls=HyperellipticCurve)
+    cls = dynamic_class(class_name, tuple(bases), doccls=HyperellipticCurve)
     return cls(PP, f, h, names=names, genus=g)
```

Comment:
You removed `HyperellipticCurve_generic` here. `HyperellipticCurve_generic` is supposed to provide the "generic" methods (from itself and from the base classes) to the fabricated class constructed by `dynamic_class`.  By your change, the "generic" methods are still accessible but placed farther in the method resolution order (mro). This does not look good.

Is it your objective to construct `your_jacobian_g2_class` object for the `HyperellipticCurve_g2` object? Then there is a standard way to solve the problem:

You add a class variable `_jacobian_class` to `HyperellipticCurve_generic` and make its `jacobian` method to construct an object of the class `_jacobian_class`. Then the subclass `HyperellipticCurve_g2` set the class variable `_jacobian_class = your_jacobian_g2`. You don't need to provide `jacobian` method to `HyperellipticCurve_g2`. 

See `sage.rings.function_field.function_field.FunctionField`, which has class variable `_differentials_space` for similar purpose.

Reply:
I understand how your fix works. 

What I suggested above only works if what you want is only `jacobian()` method.

It seems that you want to override more methods of `HyperellipticCurve_generic` in your subclass. Then the problem is normally solved by python's native inheritance mechanism of classes. Unfortunately `HyperellipticCurve` is not using the normal mechanism, but is using `dynamic_class`. 

If you want to stick to `dynamic_class` mechanism, perhaps what you did in this PR seems the only way to solve the problem.

If you want to switch to inheritance of classes, it means more work to you. You need to define more subclasses of `HyperellipticCurve_generic`. I believe this is doable. You may check `Curve` classes. 

I think `dynamic_class` was invented to solve the problem of combinatorial explosion of classes in the categories framework. It was not intended for the current case that defining a few classes in the right way solves the problem.

Reply:
OK -- well I can do the extra work if this is the only way to get this PR merged. Otherwise I believe this PR is a way to fix the current code which does not work as intended.

Reply:
I suppose here the dynamic class creation may end up being helpful if there are optimisations for many different genus -- but I can't see this happening

Reply:
> OK -- well I can do the extra work if this is the only way to get this PR merged. 

There seems to be misunderstanding. I didn't request any extra work for me to approve this PR. 

If you would want to abandon using `dynamic_class` and adopt inheritance mechanism to achieve what you want, then you would need an extra work (to define needed subclasses). That's all.

If you are okay with tweaking the `dynamic_class` mechanism already set in place  (and I guess so), then I am okay with it too.

Reply:
Sorry for the misunderstanding. 

I don't really mind what the fix is, I just want to do the correct fix. I am happy to either leave it as it is or make these additional classes.

Looking through the current usage of the dynamic class there are things like...

```
class = base_class
if condition:
    class = dynamic_class("name", (class, additional_class))
```
Which we could also do here but i dont think it's any cleaner than what we currently have

Reply:
Let's minimize changes for now.

Later you may attempt to refactor subclasses of `HyperellipticCurve_generic` to get rid of `dynamic_class` if you want for whatever reason.